fix(template-retail-react-app): product-list refinements

breadadams opened the pull request:
<!--- Provide a short summary of your changes in the Title field above -->

# Description

This PR fixes numerous points opened in #901 regarding the PLP refinements component, such as:

1. Having numerous possible types (`string | number | Array<string | number>`) for the filter values passed down to a value component (color, checkbox, etc), and then running methods like `.includes()` against them, providing different results depending on the type.
2. Invalid `isChecked` logic for checkboxes, making all checkboxes in a refinement group appear as checked after selecting just one.
3. Confusing `string|array` proptype on the color refinements component.
4. Invalid handling of query parameters, where a number is received but it was then treated as a string (by attempting to access the `.split()` method on it).

In addition to a couple of other bug fixes:

5. Not being able to deselect a "non multiple" refinement (e.g. checkbox) by unchecking the selected checkbox (as it would select itself again). This was introduced (and resolved) in this PR after having fixed item 2 from above.
6. Radio groups maintaining their previously checked UI after having their value cleared via the "Selected Refinements" clear filter buttons.

<!--- A longer summary of your changes, including: a description of the issue that you’re addressing, a list of required dependencies (if applicable), and any other relevant context. -->

# Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [x] **Bug fix** (non-breaking change that fixes an issue)
- [ ] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [ ] **Breaking change** (could cause existing functionality to not work as expected)
- [ ] **Other changes** (non-breaking changes that does not fit any of the above)

> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

_See details in commit message and description above._

# How to Test-Drive This PR

Steps for testing the (testable) points described above:

### Item 1

If you potentially had a refinement option that was a substring of another, for instance color, "blue" and "light blue". When the selected value was "light blue", both would appear as selected since here it would be using the `string.includes()` method, instead of `array.includes()`.

### Item 2

1. Visit the [electronics category](https://pwa-kit.mobify-storefront.com/us/en-US/category/electronics).
2. Select the "Type → Lithium Ion" refinement checkbox.
3. Both "Lithium Ion" and "Lithium Polymer" will be checked.

### Item 4

1. Visit the [mens suits category](https://pwa-kit.mobify-storefront.com/us/en-US/category/mens-clothing-suits).
2. Select the "Size → 36" refinement option.
3. (Note it's not marked as selected visually).
4. Select "Size → 38".
5. Nothing happens, see an error in the browser console.

### Item 6

1. Visit the [electronics category](https://pwa-kit.mobify-storefront.com/us/en-US/category/electronics).
2. Select the "Price → $20 - $49.99" refinement radio option.
3. Clear the refinement option via the individual clear buttons or the "Clear all" button in the "Selected refinements" area.
4. Notice that the radio option selected in step 2 is still visually checked.

# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [x] Changes are covered by test cases
- [ ] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [x] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

breadadams committed:
fix(template-retail-react-app): product-list refinements

+ Ensure that the `selectedFilters` provided to the `Refinements` value components is always an array. And update said value components (checkbox, color, radio & size) to handle always receiving an array.
+ Apply some loose equality checks, catering for cases where a refinement URL param is parsed as a number but the refinement `.value` is a stringified number.
+ Fix bug where non-multiple filters could not be unchecked by selecting themselves (e.g. clicking an already checked checkbox).
+ Fix bug where a refinement URL param is parsed as a number, but was handled as a string (with `.split()`).
+ Fix bug where radio refinements would still appear as checked after clearing the value via `SelectedRefinements`.
+ Deprecate unused proptypes in `SizeRefinements`

breadadams commented on packages/template-retail-react-app/app/pages/product-list/partials/radio-refinements.jsx:
```diff
@@ -12,7 +12,11 @@ import PropTypes from 'prop-types'
+            <RadioGroup
+                // The following `false` fallback is required to avoid the radio group
+                // from switching to "uncontrolled mode" when `selectedFilters` is empty.
+                value={selectedFilters[0] ?? false}
```
This is the key to fixing item 6 from the points listed in the description.

If `RadioGroup` receives undefined or null in the `value` prop, it no longer acts in controlled mode and doesn't handle the checked state of inner radio buttons.

This _could_ cause issues if a refinement had an option where `.value` is equal to `false`. In that case the `false` fallback would just need to be replaced with any other non-null value (e.g. `''`).

bendvc replied:
Nice catch.

breadadams commented on packages/template-retail-react-app/app/pages/product-list/partials/checkbox-refinements.jsx:
```diff
@@ -15,17 +15,14 @@ const CheckboxRefinements = ({filter, toggleFilter, selectedFilters}) => {
-                                        !!selectedFilters,
-                                        false
-                                    )
+                                    toggleFilter(value, filter.attributeId, isChecked, false)
```
I didn't apply it yet, but whilst making these changes on the project I'm working on I've also removed the final `false` argument (`allowMultiple`) from the `toggleFilter` method. Thus making it take the default of `true`.

This being because IMO checkboxes should represent an option that can have multiple items selected (and be deselected). Whereas currently they're essentially serving as radio buttons.

If you agree I can remove it here too.

bendvc replied:
I'm open to this based on your reasoning. But that final argument is being used used when called from the radio-refinements component. Is your change away of this and retain the correct functionality in that component?

breadadams replied:
Hey @bendvc, yeah it is, sorry my comment here wasn't very clear.

I don't mean removing the argument from the actual `toggleFilter` method, just removing the `false` argument here where it's called within `CheckboxRefinements` (so that the original method's default value is applied).

bendvc replied:
@breadadams  Ok 💯 makes sense. I'm ok with that change. Go ahead and commit that, and we can get this merged.

breadadams replied:
@bendvc, great! Resolved in latest commit 🙂

vmarta commented:
Thank you @breadadams for this PR. I plan to review it tomorrow.

Update: sorry my bad, I haven't had a chance to look at this yet. Will get back to this as soon as I can.

bendvc committed:
Merge branch 'develop' into fix/product-list-refinements

bendvc committed:
Merge branch 'develop' into fix/product-list-refinements

bendvc committed:
Merge branch 'develop' into fix/product-list-refinements

bendvc commented on packages/template-retail-react-app/app/pages/product-list/index.jsx:
```diff
@@ -195,25 +195,34 @@ const ProductList = (props) => {
             let attributeValue = searchParamsCopy.refine[attributeId] || []
-            let values = Array.isArray(attributeValue) ? attributeValue : attributeValue.split('|')
+
+            // Ensure that the value is still converted into an array if it's a `string` or `number`.
+            if (typeof attributeValue === 'string') {
+                attributeValue = attributeValue.split('|')
+            } else if (typeof attributeValue === 'number') {
+                attributeValue = [attributeValue]
+            }
```
Originally the number values in the query string were not being parsed as numbers and were being treated as strings. Looks like we broke this logic after making the change to parse numbers. 

Thanks for catching that. 👍

bendvc commented:
Hey @breadadams .. thanks for the contribution. I've added some replies to you comments. Lets work on getting that resolved and we can merged your PR 👍

breadadams committed:
feat(retail-react-app): update `CheckboxRefinements`

+ Adjust the `onChange` logic so that multiple checkbox refinements can be selected

breadadams committed:
Merge remote-tracking branch 'upstream/develop' into fix/product-list-refinements

bendvc reviewed (approved):
Hey Brad, Really appreciate the work for this fix. Everything looks a-ok 👍
